### ISSUE TITLE
SaveAsync() not working on images?

### ISSUE THREAD

**Issue by dwt12777** [AlexGhiondea/SmugMug.NET#31]
I'm very new to GitHub and a novice developer, so huge apologies if this isn't the place for this type of question.  I'm using V2 of the API and it's working great so far - I am able to read folders, albums and images from my smugmug site.  I'm also able to update folders and albums (I'm using the API to programmatically rename folders and albums), but for some reason I can't save changes to images themselves.  Specifically, I'm trying to update image captions.  The function where I'm attempting to do this is listed below.  The GetDefaultCaption() extension returns the filename of the image after stripping off the file extension and a prefix that my filenames have, which I've confirmed is working fine.  I've also confirmed (stepping through the debugger) that image.SaveAsync() is called yet nothing happens.  No error is thrown and the loop continues as if the changes are being saved.  Yet when I go back to my smugmug site the captions clearly aren't being saved.  Am I doing something wrong or does the API not yet support saving images?

```
public List<ImageEntity> UpdateImagesWithDefaultCaptions()
{
	var updatedImages = new List<ImageEntity>();

	foreach (var image in this.Images)
	{
		if (image.Caption == null || image.Caption != image.GetDefaultCaption())
		{
			image.Caption = image.GetDefaultCaption();
			image.SaveAsync();
			updatedImages.Add(image);
		}
	}
	return updatedImages;
}
```

**Comment by AlexGhiondea** [comment]
Thanks for reporting this. I wonder if the change tracking isn't working right in this scenario.

I might have some time to look into this later tonight.

**Comment by AlexGhiondea** [comment]
@dwt12777 I looked into this and what seems to happen is that I get a 401 Unauthorized response from the server when I make the patch request.

It seems like I need to add `-0` to the image as the URI for the patch is being created. I will see about creating a fix for this.

**Comment by AlexGhiondea** [comment]
Ok -- I pushed a new version that fixes the problem. Please take a look.

I am considering spending some time and polishing the API shape a lot more than it is now before the 1.0.0 release :).

Let me know if you would be willing to give it a try if I start doing it :).

**Comment by dwt12777** [comment]
Hey Alex - it's still not working on my end.  Again, keep in mind I'm pretty new to all this, so I can't promise I'm not screwing something up!  Here's what I've done:

1. Went to the Nuget package manager and told it to install 1.0.0-beta5.  When I look in my project's packages.config file I see an entry: <package id="SmugMug.Net.v2Api" version="1.0.0-beta5" targetFramework="net472" />  Is there anything else I need to do to make sure it's using the latest?
2. I re-ran my code where it loops through all images and attempts to change the caption.  The SaveAsync() call is made, but again I don't see any changes happening on the website.

Anything I can do on my end to better troubleshoot?

And yes, if you're thinking of making a push towards 1.0.0 and need some help, I'd be happy to test and provide feedback as I can.

**Comment by AlexGhiondea** [comment]
@dwt12777 thanks for giving it a try! I will keep you updated with any change to the library.

Try changing the code you have to this:
`image.SaveAsync().Wait()`

The `SaveAsync` method is async and the way you were calling it you didn't wait for the method to return. Adding the `Wait()` call makes the async method actually complete execution before you add it to the list of updated images.

**Comment by dwt12777** [comment]
Okay, I'm not sure about this, but check this out.  I noticed it's working for some of the images on my site, but not others.  When I compared the properties of an ImageEntity object that worked with an ImageEntity that didn't work, I noticed the following:

Image that worked: ```Uri = /api/v2/album/97sZCs/image/SK68Wzx-0```

Image that didn't work: ```Uri = /api/v2/album/PFGH95/image/9pHz755-1```

The suffix is different: -0 vs -1.  No idea why they're different.  Nor do I know if other suffixes are possible.  But I wonder if, rather than having to anticipate every possibility, could your code not read from the image object what the proper suffix should be?  (That is, if this even is the cause of the problem).
